### PR TITLE
Add implementation for OBJECT

### PR DESCRIPTION
**Pull request by arunvickram**
Implemented the OBJECT form. Relatively self explanatory, it basically takes all the fields of the object and the name of the object and stores it in a `Hash`

### CODE REVIEW

File: samples/99_zil_interpreter/app/builtins.rb [hunk 1]
```diff
@@ -262,3 +262,27 @@ def expect_minimum_argument_count!(arguments, min_count)
   false
 }
 
+ZIL_BUILTINS[:OBJECT] = lambda { |arguments, context|
```
Comment (leviongit) [inline samples/99_zil_interpreter/app/builtins.rb]: I think that it should eval its arguments, but that would need to be backed up by someone

File: samples/99_zil_interpreter/app/builtins.rb [hunk 2]
```diff
@@ -262,3 +262,27 @@ def expect_minimum_argument_count!(arguments, min_count)
...
+  # Objects have a name, and a list of properties and values
+
+  raise FunctionError, "OBJECT requires at least a name!" unless arguments.length > 0
```
Comment (leviongit) [inline samples/99_zil_interpreter/app/builtins.rb]: you could use my fancy helpers :wink: `expect_minimum_argument_count!(arguments, 1)` in this case

File: samples/99_zil_interpreter/app/builtins.rb [hunk 3]
```diff
@@ -262,3 +262,27 @@ def expect_minimum_argument_count!(arguments, min_count)
...
+
+  # not sure if this is needed
+  raise FunctionError, "OBJECT requires properties" unless arguments.length > 1
```
Comment (leviongit) [inline samples/99_zil_interpreter/app/builtins.rb]: I don't *think* it is?

File: samples/99_zil_interpreter/app/builtins.rb [hunk 4]
```diff
@@ -262,3 +262,27 @@ def expect_minimum_argument_count!(arguments, min_count)
...
+    raise FunctionError, "Parameters to object must be in list form" unless property.class == Syntax::List
+    raise FunctionError, "OBJECT properties require a name and values!" unless property.elements.length > 1
+
+    property_name, *property_values = property.elements
+
+    object[:properties][property_name] = property_values.length == 1 ? property_values[0] : property_values
+  end
+
+  object
+}
```
Comment (leviongit) [inline samples/99_zil_interpreter/app/builtins.rb]: this should assign the object to the global variable under it's name, not return it, thus the creation of `object` is not needed as it currently stands, it should be a bare hash (but again I may be wrong (cc) @kfischer-okarin)

**Review by leviongit — CHANGES_REQUESTED** [review]

File: samples/99_zil_interpreter/app/builtins.rb [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+  context.locals[var_atom]
+}
+
+ZIL_BUILTINS[:VALUE] = define_for_evaled_arguments { |arguments, context|
+  expect_argument_count!(arguments, 1)
+  var_atom = arguments[0]
+
+  context.locals[var_atom] || context.globals[var_atom] ||
+    (raise FunctionError, "No local nor global value for #{var_atom.inspect}")
+}
```
Comment (cbilski) [inline samples/99_zil_interpreter/app/builtins.rb]: You're going to want to back this PR out and resubmit with your HEAD at the same spot as the zil-interpreter HEAD.  We should ideally only see the two commits at the end.

File: samples/99_zil_interpreter/app/builtins.rb [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+  object_properties.each do |property|
+    raise FunctionError, "Parameters to object must be in list form" unless property.class == Syntax::List
+    raise FunctionError, "OBJECT properties require a name and values!" unless property.elements.length > 1
+
+    property_name, *property_values = property.elements
+
+    object[:properties][property_name] = property_values.length == 1 ? property_values[0] : property_values
+  end
+
+  object
```
Comment (cbilski) [inline samples/99_zil_interpreter/app/builtins.rb]: I think object needs to be saved/defined somewhere.  If we just return it, it's just going to be garbage collected, right?

File: samples/99_zil_interpreter/app/builtins.rb [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+  # not sure if this is needed
+  raise FunctionError, "OBJECT requires properties" unless arguments.length > 1
+
+  object_name, *object_properties = arguments
+
+  object = { name: object_name, properties: {} }
+
+  object_properties.each do |property|
+    raise FunctionError, "Parameters to object must be in list form" unless property.class == Syntax::List
```
Comment (cbilski) [inline samples/99_zil_interpreter/app/builtins.rb]: You'll want to look at examples in the ZIL source that contain comments and make sure your code can handle those.

```
<OBJECT TROPHY-CASE	;"first obj so L.R. desc looks right."
	(IN LIVING-ROOM)
	(SYNONYM CASE)
	(ADJECTIVE TROPHY)
	(DESC "trophy case")
	(FLAGS TRANSBIT CONTBIT NDESCBIT TRYTAKEBIT SEARCHBIT)
	(ACTION TROPHY-CASE-FCN)
	(CAPACITY 10000)>
```

Comment (cbilski) [reply samples/99_zil_interpreter/app/builtins.rb]: If you haven't, it's a good idea to look over the 1dungeon.zil code from the historical source zork1 code.  You may have already - it's hard to tell in this PR because of all of the extra commits.

File: samples/99_zil_interpreter/app/builtins.rb (hunk 5 above)
Comment (kfischer-okarin) [reply samples/99_zil_interpreter/app/builtins.rb]: I took care of that... It was just the base branch of the Pull Request that was not set correctly....

File: samples/99_zil_interpreter/app/builtins.rb (hunk 7 above)
Comment (leviongit) [reply samples/99_zil_interpreter/app/builtins.rb]: this will not handle comments I believe (I may be wrong though)

Comment (arunvickram) [reply samples/99_zil_interpreter/app/builtins.rb]: This doesn't handle comments. I was wondering if comments were ignored in the lexer/parser but looks like it doesn't. I'll change the code to adapt to it

Comment (arunvickram) [reply samples/99_zil_interpreter/app/builtins.rb]: I believe it handles comments now that it leverages `define_for_evaled_arguments`

File: readme.md [hunk 8]
```diff
@@ -68,4 +68,5 @@ Now that we are in the Pull requests tab, you can go ahead and click the nice gr
  
 You will now be in the pull request compare stage. It goes over all the changes you have made to the base code. This will take you too the comment page. You should always leave a few comments about what you changed and why. It helps people who follow you understand why you made these changes. Now hit that button (It is another green one)!
  
-Congratulations! Once an admin merges your changes, you will be a DragonRubyGTK contributor. Wear that title with pride!
\ No newline at end of file
+Congratulations! Once an admin merges your changes, you will be a DragonRubyGTK contributor. Wear that title with pride!
+
```
Comment (kfischer-okarin) [inline readme.md]: this change seems unrelated... can you undo that?

File: samples/99_zil_interpreter/tests/builtins_tests.rb [hunk 9]
```diff
@@ -578,3 +578,14 @@ def test_builtin_cond(args, assert)
...
 end
+
+def test_builtin_object(args, assert)
+  zil_context = build_zil_context(args)
+
+  # <OBJECT ROOM (HEIGHT 10)>
+  specs = [:ROOM, list(:HEIGHT, 10)]
+  result = call_routine zil_context, :OBJECT, specs
+  assert.equal! zil_context.globals[:ROOM][:name], :ROOM, "Object's name is ROOM"
+  assert.equal! zil_context.globals[:ROOM][:properties][:HEIGHT], 10, "ROOM's HEIGHT is 10"
```
Comment (kfischer-okarin) [inline samples/99_zil_interpreter/tests/builtins_tests.rb]: the messages should act as an explanation about what's went wrong... so maybe putting them as a negative statement or in a "should" statement would fit better

File: samples/99_zil_interpreter/app/builtins.rb [hunk 10]
```diff
@@ -262,3 +262,26 @@ def expect_minimum_argument_count!(arguments, min_count)
...
+
+  object = { properties: {} }
+  object[:name] = object_name
+  
+
+  object_properties.each do |property|
+    raise FunctionError, "OBJECT properties require a name and values!" unless property.length > 1
+
+    property_name, *property_values = property
+    puts property_values
```
Comment (kfischer-okarin) [inline samples/99_zil_interpreter/app/builtins.rb]: This seems to be a debug statement - so maybe remove it from the final PR ;)?

File: samples/99_zil_interpreter/app/builtins.rb (hunk 2 above)
Comment (kfischer-okarin) [reply samples/99_zil_interpreter/app/builtins.rb]: +1 to this.... :)

**Review by kfischer-okarin — CHANGES_REQUESTED** [review]
Few little comments..... Implementation looks good so far..... Once we figure out parsing and how to store defined vocabulary we probably have to adapt it a bit... since I think ADJECTIVE, SYNONYM will need to store their values in a vocabulary with a back reference to this object etc...

But well once we get there we can deal with that ;)

File: samples/99_zil_interpreter/app/builtins.rb [hunk 11]
```diff
@@ -262,3 +262,23 @@ def expect_minimum_argument_count!(arguments, min_count)
...
 }
 
+ZIL_BUILTINS[:OBJECT] = define_for_evaled_arguments { |arguments, context|
+  # Objects have a name, and a list of properties and values
+  expect_minimum_argument_count!(arguments, 1)
+  
+  object_name, *object_properties = arguments
+
+  object = { properties: {} }
+  object[:name] = object_name
```
Comment (leviongit) [inline samples/99_zil_interpreter/app/builtins.rb:272]: I *think* that the object knowing it's own name isn't required, I may be wrong though

**Review by leviongit — APPROVED** [review]

File: samples/99_zil_interpreter/app/builtins.rb (hunk 11 above)
Comment (kfischer-okarin) [reply samples/99_zil_interpreter/app/builtins.rb:272]: Hm don't know yet... but might be useful for debugging purposes.... So let's leave it in for now :D We can optimize that away later

**Review by kfischer-okarin — APPROVED** [review]
LGTM.... :)
